tools/cache_creator/cache_creator.cpp

```diff
@@ -161,6 +163,61 @@ bool hexStringToUuid(llvm::StringRef hexStr, llvm::MutableArrayRef<uint8_t> outU
   return true;
 }
 
+llvm::Error createBadPalMetadataError() {
+  return llvm::createStringError(llvm::object::object_error::invalid_file_type, "Unexpected PAL metadata format");
+}
+
+// =====================================================================================================================
+// Tries to extract cache hash and LLPC version from a PAL metadata note blob.
+//
+// @param noteBlob : The input metadata blob
+// @returns : The cache info on success, or error on failure
+llvm::Expected<ElfLlpcCacheInfo> getCacheInfoFromMetadataBlob(llvm::StringRef noteBlob) {
+  llvm::msgpack::Document document;
+  if (!document.readFromBlob(noteBlob, false))
+    return createBadPalMetadataError();
+
+  // Traverse document to XGL cache info node.
+  llvm::msgpack::DocNode node = document.getRoot();
+  if (!node.isMap())
+    return createBadPalMetadataError();
+  node = node.getMap(false)["amdpal.pipelines"];
+  if (!node.isArray())
+    return createBadPalMetadataError();
+  node = node.getArray(false)[0];
```

Comment:
If `node` is not an array, the `getArray(false)` will cause an assert failure.  It will extend the array with empty nodes if the element does not exist, and return the empty node (https://llvm.org/doxygen/classllvm_1_1msgpack_1_1ArrayDocNode.html#a9df88e85c1ec39c1604e13c22e38f3ef).

Reply:
Right, if the key doesn't exist in the map/array, a node with  `Empty` type will be returned and the following `isX` check will return false.